core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -270,6 +271,33 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     currentStateTimestamp = Some(time.milliseconds())
   }
 
+  /**
+    * Shrinks the consumer group to the given size by removing members while maintaining the leader
+    */
+  def shrinkTo(maxSize: Int): Unit = {
+    if (members.size < maxSize)
+      throw new IllegalArgumentException(s"Cannot shrink group $groupId to $maxSize as it's current count ${members.size} is smaller.")
+
+    var leaderWasAdded = leaderId.isEmpty
+    val newMembers = new mutable.HashMap[String, MemberMetadata]
+    val maxSizeWithoutLeader = maxSize - 1
+
+    members.foreach(memberEntry => {
```

Comment:
Yeah this is pretty inefficient.
My thinking is that we should run through this code path only on broker rolls with a new config. But regarding your other comment, shrinking may turn out to be completely unnecessary. I'll wait for @hachikuji to give his thoughts as well before removing it.

Reply:
I agree that @gwenshap's idea is more efficient. Otherwise, if we are sticking to the existing logic, we might be able to leverage the `retain(...)` function and avoid creating a second hash map.

Reply:
`retain()` is a cool function but I will go with @gwenshap's suggestion as it makes the updating of `supportedProtocols` easier